packages/dds/map/src/mapKernel.ts

```diff
@@ -112,6 +111,21 @@ export interface IMapDataObjectSerialized {
     [key: string]: ISerializedValue;
 }
 
+type PreviousMap = Map<string, ILocalValue> | undefined;
+
+interface IMapClearLocalOpMetadata {
+    pendingMessageId: number;
+    previousMap: PreviousMap;
+}
+
+// "none" means the key was not set before the edit. undefined means the previous value is unknown.
+type PreviousKeyValue = ILocalValue | "none" | undefined;
+
+interface IMapKeyLocalOpMetadata {
+    pendingMessageId: number;
+    previousValue: PreviousKeyValue;
+}
+
```

Comment:
i'm having trouble following the typing of localOpMetadata expecially with all the casting. I think we should create a union type, `const type MapLocalMetadata = IMapKeyLocalOpMetadata | IMapClearLocalOpMetadata` and use that everywhere. also rather than `as` casting  which isn't safe use a discrimated union by adding the op typ to each local metadata type

Reply:
i think in most places in this file you will be able to avoid casting. there maybe a couple places near entry points where you need to do an inital case from unknown too MapLocalMetadata, but from then on you shouldn't need to. you could probably even get ride of that with type guard, `const isMapClearMetadata = (lm: any): lm is MapClearLocalOpMetadata=> lm has "pendingMessageId" &&  lm.type === "clear"

Reply:
I'm not sure it's much better now. Let me know if it's sufficiently comprehensible.